### PR TITLE
Added Hackathon endpoints

### PR DESCRIPTION
**Pull request by aishwary023**
Added:
- Hackthon List 
- Hackthon Teams list
- Hackthon Team create
endpoints.

### CODE REVIEW

**Comment by aishwary023** [comment]
@nishantwrp Resolved conflicts, PTAL, thanks

File: core/views.py [hunk 1]
```diff
@@ -1,10 +1,64 @@
-from rest_framework import permissions
-from rest_framework import generics
-from rest_framework import status
-from .serializers import HackathonSerializer
+from rest_framework import generics, status, permissions
+from rest_framework.response import Response
+from django.utils import timezone
+import pytz
```
Comment (nishantwrp) [inline core/views.py]: Why this?

File: core/views.py [hunk 2]
```diff
@@ -1,10 +1,64 @@
...
+    completed:
+    List of completed Hackathons
+    """
+    serializer_class = HackathonSerializer
+    permission_classes = [permissions.AllowAny]
+
+    def get_queryset(self):
+        queryset = Hackathon.objects.all()
+        query = self.request.query_params.get('query', None)
+        current_date = timezone.now()
```
Comment (nishantwrp) [inline core/views.py:94]: This depends on the time zone of the server where this app will be deployed, right?

File: core/urls.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 from django.urls import path
-from .views import (HackathonsRUDView,HackathonCreateView)
+from .views import HackathonTeamListView, HackathonTeamCreateView, HackathonListView, HackathonsRUDView, HackathonCreateView
 
-urlpatterns = [   
+urlpatterns = [
+    path('hackathons/<int:pk>/teams/', HackathonTeamListView.as_view()),
+    path('hackathons/add-team/', HackathonTeamCreateView.as_view()),
```
Comment (nishantwrp) [inline core/urls.py]: This can be shifted to POST hackathons/<int:pk>/teams/

File: core/views.py [hunk 4]
```diff
@@ -1,10 +1,64 @@
...
+        if query is not None:
+            if(query == 'ongoing'):
+                queryset = Hackathon.objects.filter(start__lt=current_date, end__gt=current_date)
+            elif(query == 'completed'):
+                queryset = Hackathon.objects.filter(start__lt=current_date, end__lt=current_date)
+            elif(query == 'upcoming'):
+                queryset = Hackathon.objects.filter(start__gt=current_date, end__gt=current_date)
+            else:
+                return Response("Invalid query parameter", status.HTTP_400_BAD_REQUEST)
+        return queryset
```
Comment (nishantwrp) [inline core/views.py]: I'm not sure but is there a way of showing the query parameters in the swagger view?

**Review by nishantwrp — CHANGES_REQUESTED** [review]
Some comments @aishwary023, ptal!

File: core/views.py (hunk 2 above)
Comment (aishwary023) [reply core/views.py:94]: timezone.now() returns the UTC date/time( dates stored in database are also in UTC) . So, it would not depend on the location of the server.

File: core/views.py (hunk 1 above)
Comment (aishwary023) [reply core/views.py]: Forgot to remove pytz, it's not required. Removed now

File: core/urls.py (hunk 3 above)
Comment (aishwary023) [reply core/urls.py]: @nishantwrp 
There is no RetrieveCreateAPIView to apply post and get in a generic view. And the post() method does not expect a 'pk' keyword. 
Should I move this to  ```POST /hackathons/teams/``` or is there any other way?

**Comment by nishantwrp** [comment]
Oh wait, Make it GET /hackathons/id/teams/create then?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: core/views.py (hunk 4 above)
Comment (aishwary023) [reply core/views.py]: Done

**Comment by aishwary023** [comment]
@nishantwrp Made the changes, PTAL

File: core/permissions.py [hunk 5]
```diff
@@ -1,12 +1,27 @@
...
\ No newline at end of file
+
+class isProfileComplete(permissions.BasePermission):
+    """
+    1. GET can be accessed even by anonymous users
+    2. POST can only be accessed by users whose profile is completed
+    """
+    def has_permission(self, request, view):
+        if request.method == 'GET':
+            return True
```
Comment (nishantwrp) [inline core/permissions.py]: This should be a general permission. So, maybe remove this,

File: core/views.py (hunk 2 above)
Comment (nishantwrp) [reply core/views.py:94]: Ohk

File: core/permissions.py (hunk 5 above)
Comment (aishwary023) [reply core/permissions.py]: done

File: core/migrations/0002_auto_20201017_0554.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.1.2 on 2020-10-16 19:01
+# Generated by Django 3.1.2 on 2020-10-17 05:54
```
Comment (nishantwrp) [inline core/migrations/0002_auto_20201017_0554.py]: Revert this change

File: core/views.py [hunk 7]
```diff
@@ -1,22 +1,88 @@
...
+        try:
+            if Hackathon.objects.get(id=kwargs['pk']) is not None:
+                queryset = Team.objects.filter(hackathon=kwargs['pk'])
+        except:
+            return Response("Hackathon does not exist!", status=status.HTTP_404_NOT_FOUND)
+        serializer = TeamSerializer(queryset, many=True)
+        return Response(serializer.data)
+
+query_param = openapi.Parameter('query', openapi.IN_QUERY, description="Query parameter - 'upcoming', 'ongoing' or 'completed'",
+                                type=openapi.TYPE_STRING)
```
Comment (nishantwrp) [inline core/views.py]:
```suggestion
query_param = openapi.Parameter(
    'query', openapi.IN_QUERY, description="Query parameter - Returns all hackthons if not specified.",
    type=openapi.TYPE_STRING, enum=['completed', 'upcoming', 'ongoing'])
```

File: core/permissions.py [hunk 8]
```diff
@@ -1,12 +1,25 @@
...
-        '''
+        """
         def has_permission(self, request, view):
             if request.method in permissions.SAFE_METHODS:
                 return True
             return bool(request.user and request.user.is_superuser)
-    
\ No newline at end of file
+
+class isProfileComplete(permissions.BasePermission):
```
Comment (nishantwrp) [inline core/permissions.py]:
```suggestion
class AllowCompleteProfile(permissions.BasePermission):
```

File: core/views.py [hunk 9]
```diff
@@ -1,22 +1,88 @@
...
+        """
+        Returns list of team registered for a hackathon
+        """
+        try:
+            if Hackathon.objects.get(id=kwargs['pk']) is not None:
+                queryset = Team.objects.filter(hackathon=kwargs['pk'])
+        except:
+            return Response("Hackathon does not exist!", status=status.HTTP_404_NOT_FOUND)
+        serializer = TeamSerializer(queryset, many=True)
+        return Response(serializer.data)
```
Comment (nishantwrp) [inline core/views.py]: You can write it in a better way. Maybe refer https://github.com/codefest-iit-bhu/codefest-app/blob/0e7c8f737ca684797ed512931870094c9b547ab7/website/views.py#L92

**Review by nishantwrp — CHANGES_REQUESTED** [review]

File: core/serializers.py [hunk 10]
```diff
@@ -1,8 +1,25 @@
...
+class TeamCreateSerializer(serializers.ModelSerializer):
+
+    def validate(self, data):
+        hackathon_id = self.context.get('view').kwargs['pk']
+        if hackathon_id != data['hackathon']:
+            raise serializers.ValidationError("Mismatch of Hackathon ID in url and request body!")
+        return data
+    class Meta:
+        model = Team
+        fields = '__all__'
```
Comment (nishantwrp) [inline core/serializers.py]: I think for creating a team you need only the team name in the request body.

File: core/views.py [hunk 11]
```diff
@@ -1,22 +1,72 @@
...
+        if self.request.method == 'GET':
+            return TeamSerializer
+        else:
+            return TeamCreateSerializer
+
+    def get_queryset(self, **kwargs):
+        if getattr(self, 'swagger_fake_view', False):
+            return None
+        try:
+            return Hackathon.objects.get(id=self.kwargs['pk'])
```
Comment (nishantwrp) [inline core/views.py]: why return the hackathon?

File: core/views.py [hunk 12]
```diff
@@ -1,22 +1,72 @@
...
+        current_date = timezone.now()
+        if query is not None:
+            if(query == 'ongoing'):
+                queryset = Hackathon.objects.filter(start__lt=current_date, end__gt=current_date)
+            elif(query == 'completed'):
+                queryset = Hackathon.objects.filter(start__lt=current_date, end__lt=current_date)
+            elif(query == 'upcoming'):
+                queryset = Hackathon.objects.filter(start__gt=current_date, end__gt=current_date)
+            else:
+                raise exceptions.ValidationError("Invalid query parameter!")
```
Comment (nishantwrp) [inline core/views.py:103]: Maybe add this logic in the serializer

**Review by nishantwrp — CHANGES_REQUESTED** [review]
Some comments, ptal!

File: core/views.py (hunk 11 above)
Comment (aishwary023) [reply core/views.py]: To check if the hackathon with that id exists or not, although instead of ```return``` it should just be
```hackathon = Hackathon.objects.get(id=self.kwargs['pk'])```

File: core/serializers.py (hunk 10 above)
Comment (aishwary023) [reply core/serializers.py]: And what about the member details, leader etc?

Comment (nishantwrp) [reply core/serializers.py]: That can be deduced from the token auth, right? And members join afterward by entering the team code.

Comment (aishwary023) [reply core/serializers.py]: Ah, alright. So, initally when team is created, we would set the logged in user as the Leader and member. And set a TeamID, and register the team in the hackathon.

Now another user can enter TeamID to join the team. And for that we'll have to create another endpoint. Should I create that endpoint too or just proceed with the initial team creation?

**Comment by nishantwrp** [comment]
Yup create another endpoint /hackathons/{id}/teams/join/{code}?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by aishwary023** [comment]
@nishantwrp ping

File: core/views.py [hunk 13]
```diff
@@ -1,22 +1,113 @@
...
+        else:
+            return [permissions.IsAdminUser()]
+
+    def get_queryset(self):
+        queryset = Hackathon.objects.all()
+        query = self.request.query_params.get('query', None)
+        current_date = timezone.now()
+        if query is not None:
+            if(query == 'ongoing'):
+                queryset = Hackathon.objects.filter(start__lt=current_date, end__gt=current_date)
```
Comment (krashish8) [inline core/views.py:97]:
```suggestion
                queryset = Hackathon.objects.filter(start__lte=current_date, end__gte=current_date)
```
A single-day hackathon might get excluded in 'ongoing'

File: core/views.py [hunk 14]
```diff
@@ -1,22 +1,113 @@
...
+    """
+    patch:
+    Join a team with team_id and hackathon_id
+    """
+    serializer_class = JoinTeamSerializer
+
+    def get_queryset(self):
+        if getattr(self, 'swagger_fake_view', False):
+            return None
+        pass
```
Comment (krashish8) [inline core/views.py:58]:
```suggestion
        return super().get_object()
```
Shouldn't we call the parent class function here?
/cc @nishantwrp

File: core/serializers.py [hunk 15]
```diff
@@ -1,8 +1,64 @@
...
+
+class TeamCreateSerializer(serializers.ModelSerializer):
+
+    def validate(self, attrs):
+        name = attrs['name']
+        hackathon_id = self.context['kwargs']['pk']
+        try:
+            hackathon = Hackathon.objects.get(id=hackathon_id)
+        except Hackathon.DoesNotExist:
+            raise exceptions.NotFound(detail="Hackathon does not exist!")
```
Comment (krashish8) [inline core/serializers.py:19]: I see that you are catching this exception in several places. This is fine, just a small suggestion - you could have used `get_object_or_404()` instead.

**Review by nishantwrp — APPROVED** [review]

**Review by krashish8 — COMMENTED** [review]
Although this PR is merged, here are some comments/suggestions from my side.

File: core/views.py (hunk 13 above)
Comment (aishwary023) [reply core/views.py:97]: Correct, I missed this. Will change

File: core/serializers.py (hunk 15 above)
Comment (aishwary023) [reply core/serializers.py:19]: Oh alright, I'll create an issue for the changes
